File: server/insar-processor.ts
/**
 * InSAR Processing Engine
 * Handles all InSAR data processing workflows including:
 * - Data download (Sentinel-1, DEM, orbit data)
 * - Coregistration
 * - Interferogram generation
 * - Phase unwrapping
 * - Deformation inversion
 */

import * as db from "./db";

export interface ProcessingConfig {
  projectId: number;
  startDate: string;
  endDate: string;
  satellite: string;
  orbitDirection: "ascending" | "descending";
  polarization: string;
  coherenceThreshold: number;
  outputResolution: number;
}

export interface ProcessingResult {
  type: "interferogram" | "coherence" | "deformation" | "dem" | "unwrapped_phase" | "los_displacement";
  fileUrl: string;
  fileName: string;
  metadata: Record<string, any>;
}

/**
 * Main processing pipeline orchestrator
 */
export class InSARProcessor {
  private projectId: number;
  private config: ProcessingConfig;

  constructor(config: ProcessingConfig) {
    this.projectId = config.projectId;
    this.config = config;
  }

  /**
   * Execute the complete InSAR processing workflow
   */
  async processProject(): Promise<void> {
    try {
      // Step 1: Data Download
      await this.executeStep("data_download", () => this.downloadData());

      // Step 2: Coregistration
      await this.executeStep("coregistration", () => this.coregister());

      // Step 3: Interferogram Generation
      await this.executeStep("interferogram_generation", () => this.generateInterferogram());

      // Step 4: Phase Unwrapping
      await this.executeStep("phase_unwrapping", () => this.unwrapPhase());

      // Step 5: Deformation Inversion
      await this.executeStep("deformation_inversion", () => this.invertDeformation());

      // Mark project as completed
      await db.updateProject(this.projectId, {
        status: "completed",
        progress: 100,
      });

      await this.log("info", "Processing completed successfully");
    } catch (error) {
      await db.updateProject(this.projectId, {
        status: "failed",
      });

      const errorMessage = error instanceof Error ? error.message : String(error);
      await this.log("error", `Processing failed: ${errorMessage}`);
      throw error;
    }
  }

  /**
   * Execute a processing step with error handling and logging
   */
  private async executeStep(
    stepName: string,
    processor: () => Promise<void>
  ): Promise<void> {
    const startTime = new Date();
    await this.log("info", `Starting ${stepName}...`);

    try {
      // Create step record
      const stepId = await db.createProcessingStep({
        projectId: this.projectId,
        stepName,
        status: "processing",
        progress: 0,
      });

      // Execute processing
      await processor();

      // Update step as completed
      const duration = Math.floor((Date.now() - startTime.getTime()) / 1000);
      await db.updateProcessingStep(stepId, {
        status: "completed",
        progress: 100,
        duration,
      });

      await this.log("info", `${stepName} completed in ${duration}s`);

      // Update overall progress
      const stepCount = 5; // Total processing steps
      const completedSteps = await this.getCompletedStepsCount();
      const progress = Math.floor((completedSteps / stepCount) * 100);
      await db.updateProject(this.projectId, { progress });
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error);
      await db.updateProcessingStep(
        (await db.getProjectSteps(this.projectId)).find((s) => s.stepName === stepName)?.id || 0,
        {
          status: "failed",
          errorMessage,
        }
      );

      await this.log("error", `${stepName} failed: ${errorMessage}`);
      throw error;
    }
  }

  /**
   * Step 1: Download Sentinel-1 data, DEM, and orbit information
   */
  private async downloadData(): Promise<void> {
    // Simulate data download
    // In production, this would:
    // 1. Query Copernicus Data Hub for Sentinel-1 products
    // 2. Download SLC/GRD products
    // 3. Download SRTM/ASTER DEM
    // 4. Download precise orbit ephemeris

    await this.log("debug", `Downloading Sentinel-1 data for ${this.config.startDate} to ${this.config.endDate}`);

    // Simulate processing time
    await new Promise((resolve) => setTimeout(resolve, 2000));

    await this.log("info", "Downloaded 2 Sentinel-1 SLC products");
    await this.log("info", "Downloaded SRTM DEM (30m resolution)");
    await this.log("info", "Downloaded precise orbit ephemeris");
  }

  /**
   * Step 2: Coregister SAR images to a common reference
   */
  private async coregister(): Promise<void> {
    await this.log("debug", "Starting coregistration process");

    // Simulate coregistration
    await new Promise((resolve) => setTimeout(resolve, 3000));

    await this.log("info", "Computed coregistration offsets");
    await this.log("info", "Applied resampling to slave image");
    await this.log("info", "Coregistration completed with RMS error: 0.05 pixels");
  }

  /**
   * Step 3: Generate interferogram from coregistered images
   */
  private async generateInterferogram(): Promise<void> {
    await this.log("debug", "Generating interferogram");

    // Simulate interferogram generation
    await new Promise((resolve) => setTimeout(resolve, 2500));

    await this.log("info", "Computed complex interferogram");
    await this.log("info", "Applied multilooking (4x4)");
    await this.log("info", "Computed coherence map");

    // Create mock result
    const result: ProcessingResult = {
      type: "interferogram",
      fileUrl: "https://example.com/results/interferogram.tif",
      fileName: "interferogram.tif",
      metadata: {
        wavelength: 0.0554,
        baseline: 125.5,
        incidenceAngle: 39.2,
      },
    };

    await db.createProcessingResult({
      projectId: this.projectId,
      resultType: "interferogram",
      fileUrl: result.fileUrl,
      fileName: result.fileName,
      format: "GeoTIFF",
      metadata: JSON.stringify(result.metadata),
    });

    await this.log("info", "Interferogram generated successfully");
  }

  /**
   * Step 4: Unwrap phase using minimum cost flow algorithm
   */
  private async unwrapPhase(): Promise<void> {
    await this.log("debug", "Starting phase unwrapping");

    // Simulate phase unwrapping
    await new Promise((resolve) => setTimeout(resolve, 3500));

    await this.log("info", "Applied phase filtering");
    await this.log("info", "Detected phase discontinuities");
    await this.log("info", "Running minimum cost flow algorithm");
    await this.log("info", "Phase unwrapping completed");

    const result: ProcessingResult = {
      type: "unwrapped_phase",
      fileUrl: "https://example.com/results/unwrapped_phase.tif",
      fileName: "unwrapped_phase.tif",
      metadata: {
        unwrappingAlgorithm: "MCF",
        residues: 42,
      },
    };

    await db.createProcessingResult({
      projectId: this.projectId,
      resultType: "unwrapped_phase",
      fileUrl: result.fileUrl,
      fileName: result.fileName,
      format: "GeoTIFF",
      metadata: JSON.stringify(result.metadata),
    });
  }

  /**
   * Step 5: Invert unwrapped phase to deformation
   */
  private async invertDeformation(): Promise<void> {
    await this.log("debug", "Starting deformation inversion");

    // Simulate deformation inversion
    await new Promise((resolve) => setTimeout(resolve, 2000));

    await this.log("info", "Converted phase to range change");
    await this.log("info", "Projected to LOS direction");
    await this.log("info", "Applied atmospheric correction");

    const result: ProcessingResult = {
      type: "los_displacement",
      fileUrl: "https://example.com/results/los_displacement.tif",
      fileName: "los_displacement.tif",
      metadata: {
        unit: "mm",
        minValue: -45.2,
        maxValue: 38.7,
        meanValue: -2.1,
      },
    };

    await db.createProcessingResult({
      projectId: this.projectId,
      resultType: "los_displacement",
      fileUrl: result.fileUrl,
      fileName: result.fileName,
      format: "GeoTIFF",
      minValue: "-45.2",
      maxValue: "38.7",
      meanValue: "-2.1",
      metadata: JSON.stringify(result.metadata),
    });

    await this.log("info", "Deformation inversion completed");
  }

  /**
   * Log a message to the processing logs
   */
  private async log(level: "debug" | "info" | "warning" | "error", message: string): Promise<void> {
    await db.addProcessingLog({
      projectId: this.projectId,
      logLevel: level,
      message,
    });
  }

  /**
   * Get count of completed processing steps
   */
  private async getCompletedStepsCount(): Promise<number> {
    const steps = await db.getProjectSteps(this.projectId);
    return steps.filter((s) => s.status === "completed").length;
  }
}

/**
 * Start processing a project
 */
export async function startProcessing(config: ProcessingConfig): Promise<void> {
  const processor = new InSARProcessor(config);
  await processor.processProject();
}

File: python_service/insar_processor.py
"""
InSAR Processing Engine - Updated with Atmospheric Correction
Main orchestrator for InSAR processing workflow
"""

import logging
from typing import Dict, Any, Optional
from pathlib import Path
from datetime import datetime
import json

logger = logging.getLogger(__name__)


class InSARProcessingEngine:
    """
    Main InSAR processing engine
    Orchestrates the complete workflow from data download to deformation inversion
    """
    
    def __init__(self, task_id: str, request):
        self.task_id = task_id
        self.request = request
        self.project_id = request.project_id
        
        # Processing steps
        self.steps = [
            "data_download",
            "dem_download",
            "coregistration",
            "interferogram_generation",
            "phase_unwrapping",
            "atmospheric_correction",
            "deformation_inversion"
        ]
        
        # Output directory
        self.output_dir = Path(f"./data/projects/{self.project_id}")
        self.output_dir.mkdir(parents=True, exist_ok=True)
        
        # Metadata
        self.metadata = {
            "task_id": task_id,
            "project_id": self.project_id,
            "start_date": request.start_date,
            "end_date": request.end_date,
            "satellite": request.satellite,
            "orbit_direction": request.orbit_direction,
            "polarization": request.polarization,
            "coherence_threshold": request.coherence_threshold,
            "output_resolution": request.output_resolution,
            "start_time": datetime.now().isoformat(),
            "steps": []
        }
        
        logger.info(f"InSARProcessingEngine initialized for task {task_id}, project {self.project_id}")
    
    def get_processing_steps(self) -> list:
        """Get list of processing steps"""
        return self.steps
    
    def get_step_description(self, step: str) -> Dict[str, str]:
        """Get description of a processing step"""
        descriptions = {
            "data_download": {
                "name": "Data Download",
                "description": "Downloading Sentinel-1 SLC products from Copernicus Data Hub",
                "inputs": ["Date range", "Orbit direction", "Polarization"],
                "outputs": ["SLC products", "Orbit files"]
            },
            "dem_download": {
                "name": "DEM Download",
                "description": "Downloading SRTM 30m digital elevation model",
                "inputs": ["AOI bounds"],
                "outputs": ["DEM GeoTIFF"]
            },
            "coregistration": {
                "name": "Coregistration",
                "description": "Coregistering SAR images to common reference geometry",
                "inputs": ["SLC products"],
                "outputs": ["Coregistered SLC images"]
            },
            "interferogram_generation": {
                "name": "Interferogram Generation",
                "description": "Generating interferogram and coherence map",
                "inputs": ["Coregistered SLC", "DEM"],
                "outputs": ["Interferogram", "Coherence map"]
            },
            "phase_unwrapping": {
                "name": "Phase Unwrapping",
                "description": "Unwrapping interferometric phase using MCF algorithm",
                "inputs": ["Interferogram"],
                "outputs": ["Unwrapped phase"]
            },
            "atmospheric_correction": {
                "name": "Atmospheric Correction",
                "description": "Estimating and removing atmospheric phase delays using DEM-based methods",
                "inputs": ["Interferogram", "DEM"],
                "outputs": ["Corrected interferogram", "APS estimates"]
            },
            "deformation_inversion": {
                "name": "Deformation Inversion",
                "description": "Converting unwrapped phase to LOS deformation",
                "inputs": ["Unwrapped phase"],
                "outputs": ["Deformation map"]
            }
        }
        
        return descriptions.get(step, {})
    
    def save_metadata(self):
        """Save processing metadata to JSON"""
        metadata_file = self.output_dir / "metadata.json"
        
        with open(metadata_file, 'w') as f:
            json.dump(self.metadata, f, indent=2)
        
        logger.info(f"Metadata saved to {metadata_file}")
    
    def add_step_result(self, step: str, result: Dict[str, Any]):
        """Add result for a processing step"""
        self.metadata["steps"].append({
            "name": step,
            "timestamp": datetime.now().isoformat(),
            "result": result
        })
        
        logger.info(f"Step {step} result recorded")
    
    def get_summary(self) -> Dict[str, Any]:
        """Get processing summary"""
        return {
            "task_id": self.task_id,
            "project_id": self.project_id,
            "total_steps": len(self.steps),
            "completed_steps": len(self.metadata["steps"]),
            "start_time": self.metadata["start_time"],
            "end_time": datetime.now().isoformat(),
            "steps": self.metadata["steps"]
        }


class ProcessingConfig:
    """Configuration for InSAR processing"""
    
    def __init__(self, **kwargs):
        self.project_id = kwargs.get("project_id")
        self.start_date = kwargs.get("start_date")
        self.end_date = kwargs.get("end_date")
        self.satellite = kwargs.get("satellite", "Sentinel-1")
        self.orbit_direction = kwargs.get("orbit_direction", "ascending")
        self.polarization = kwargs.get("polarization", "VV")
        self.aoi_bounds = kwargs.get("aoi_bounds")
        self.coherence_threshold = kwargs.get("coherence_threshold", 0.4)
        self.output_resolution = kwargs.get("output_resolution", 30)
        
        # Validation
        self._validate()
    
    def _validate(self):
        """Validate configuration"""
        if not self.project_id:
            raise ValueError("project_id is required")
        
        if not self.start_date or not self.end_date:
            raise ValueError("start_date and end_date are required")
        
        if self.orbit_direction not in ["ascending", "descending"]:
            raise ValueError("orbit_direction must be 'ascending' or 'descending'")
        
        if not 0 <= self.coherence_threshold <= 1:
            raise ValueError("coherence_threshold must be between 0 and 1")
        
        logger.info("Configuration validation passed")
    
    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary"""
        return {
            "project_id": self.project_id,
            "start_date": self.start_date,
            "end_date": self.end_date,
            "satellite": self.satellite,
            "orbit_direction": self.orbit_direction,
            "polarization": self.polarization,
            "aoi_bounds": self.aoi_bounds,
            "coherence_threshold": self.coherence_threshold,
            "output_resolution": self.output_resolution
        }


class ProcessingResult:
    """Result of InSAR processing"""
    
    def __init__(self, task_id: str, project_id: int):
        self.task_id = task_id
        self.project_id = project_id
        self.status = "pending"
        self.results = {}
        self.errors = []
        self.logs = []
        self.start_time = datetime.now()
        self.end_time = None
    
    def set_success(self):
        """Mark processing as successful"""
        self.status = "completed"
        self.end_time = datetime.now()
        logger.info(f"Processing {self.task_id} completed successfully")
    
    def set_error(self, error: str):
        """Mark processing as failed"""
        self.status = "failed"
        self.end_time = datetime.now()
        self.errors.append(error)
        logger.error(f"Processing {self.task_id} failed: {error}")
    
    def add_result(self, key: str, value: Any):
        """Add a result"""
        self.results[key] = value
    
    def add_log(self, level: str, message: str):
        """Add a log entry"""
        self.logs.append({
            "timestamp": datetime.now().isoformat(),
            "level": level,
            "message": message
        })
    
    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary"""
        return {
            "task_id": self.task_id,
            "project_id": self.project_id,
            "status": self.status,
            "results": self.results,
            "errors": self.errors,
            "logs": self.logs,
            "start_time": self.start_time.isoformat(),
            "end_time": self.end_time.isoformat() if self.end_time else None,
            "duration_seconds": (self.end_time - self.start_time).total_seconds() if self.end_time else None
        }
